src/types/Firestore.ts

```diff
@@ -1,88 +1,90 @@
-import { AdminRole } from "hooks/roles";
+import { Table } from "./Table";
+import { User } from "./User";
+import { Venue } from "./Venue";
+import { VenueEvent } from "./VenueEvent";
 
-interface VenueStatus {
-  currentVenue: boolean;
-  currentEvent: boolean;
-  eventPurchase: boolean;
-  venueChats: boolean;
-  venueEvents: boolean;
-  userPurchaseHistory: boolean;
-}
+export type AnyVenue = Venue | PartyMapVenue | CampVenue;
 
 interface Experience {
   reactions: Record<string, Reaction>;
   tables: Record<string, Table>;
 }
 
-type VenueTimestamps = Record<keyof VenueStatus, number>;
-export type AnyVenue = Venue | PartyMapVenue | CampVenue;
-
 interface UserVisit {
   timeSpent: number;
 }
 
+export type ValidFirestoreKeys = keyof FirestoreData | keyof FirestoreOrdered;
+
 export interface Firestore {
-  status: {
-    requesting: VenueStatus;
-    requested: VenueStatus;
-    timestamps: VenueTimestamps;
-  };
-  data: {
-    currentVenue?: AnyVenue;
-    parentVenue?: AnyVenue;
-    currentEvent: Record<string, VenueEvent>;
-    venueChats: Record<string, RestrictedChatMessage> | null;
-    venueEvents: Record<string, VenueEvent>;
-    userPurchaseHistory: Record<string, Purchase>;
-    partygoers: Record<string, User>;
-    users: Record<string, User>;
-    privatechats: Record<string, PrivateChatMessage>;
-    experiences: Record<string, Experience>;
-    eventPurchase: Record<string, Purchase>;
-    reactions: Record<string, Reaction>;
-    venues?: Record<string, AnyVenue>;
-    events?: Record<string, VenueEvent>;
-    playaVenues?: Record<string, AnyVenue>; // for the admin playa preview
-    allUsers?: Record<string, User>;
-    userModalVisits?: Record<string, UserVisit>;
-    userRoles: Record<string, Role>;
-    allowAllRoles: Record<string, Role>;
-    adminRole: AdminRole;
-  };
-  ordered: {
-    currentVenue: Array<WithId<AnyVenue>>;
-    currentEvent: Array<WithId<VenueEvent>>;
-    venueChats: Array<WithId<RestrictedChatMessage>>;
-    venueEvents: Array<WithId<VenueEvent>>;
-    userPurchaseHistory: Array<WithId<Purchase>>;
-    partygoers: Array<WithId<User>>;
-    users: Array<WithId<User>>;
-    privatechats: Array<WithId<PrivateChatMessage>>;
-    experiences: Array<WithId<Experience>>;
-    eventPurchase: Array<WithId<Purchase>>;
-    reactions: Array<WithId<Reaction>>;
-    venues?: Array<WithId<AnyVenue>>;
-    events?: Array<WithId<VenueEvent>>;
-    playaVenues?: Array<WithId<AnyVenue>>;
-    statsOnlineUsers?: Array<WithId<User>>;
-    statsOpenVenues?: Array<WithId<AnyVenue>>;
-    allUsers?: Array<WithId<User>>;
-    userModalVisits?: Array<WithId<UserVisit>>;
-    chatRequests?: Array<WithId<ChatRequest>>;
-  };
+  data: FirestoreData;
+  ordered: FirestoreOrdered;
+  status: FirestoreStatus;
+}
+
+export interface FirestoreStatus {
+  requesting: Record<ValidFirestoreKeys, boolean>;
+  requested: Record<ValidFirestoreKeys, boolean>;
+  timestamps: Record<ValidFirestoreKeys, number>;
+}
```

Comment:
Note that these types are now automagically constructed for us, so long as we correctly add an entry to both `FirestoreData` and `FirestoreOrdered`.